### PR TITLE
Fix type of call to forName

### PR DESCRIPTION
**Pull request by NathanJPhillips**

### CODE REVIEW

**Review by smowton — APPROVED** [review]

**Comment by smowton** [comment]
@allredj @romainbrenguier probably well placed to approve

**Review by romainbrenguier — APPROVED** [review]

File: src/java_bytecode/java_string_library_preprocess.cpp [hunk 1]
```diff
@@ -1517,7 +1517,8 @@ codet java_string_library_preprocesst::make_object_get_class_code(
   fun_call.lhs()=class1;
   fun_call.arguments().push_back(string1);
   code_typet fun_type;
-  fun_type.return_type()=string1.type();
+  fun_type.parameters().push_back(code_typet::parametert(string_ptr_type));
+  fun_type.return_type()=class_type;
   fun_call.function().type()=fun_type;
```
Comment (tautschnig) [inline src/java_bytecode/java_string_library_preprocess.cpp:1582]: I'd suggest to re-arrange the code so as to set up that type before creating the `symbol_exprt` above.

**Review by thk123 — APPROVED** [review]
It'd be great to see a regression test that would exhibit the problem this caused, or a unit test. The change looks correct.

**Review by allredj — APPROVED** [review]

**Comment by tautschnig** [comment]
@NathanJPhillips would it be possible to add a test as suggested by @thk123?

File: unit/testing-utils/load_java_class.cpp [hunk 2]
```diff
@@ -43,6 +45,8 @@ symbol_tablet load_java_class(
   // Configure the path loading
   cmdlinet command_line;
   command_line.set("java-cp-include-files", class_path);
+  if(use_string_refinement)
```
Comment (thk123) [inline unit/testing-utils/load_java_class.cpp]: This doesn't actually work - `cmdlinet` will ignore all calls to `set` if the command line parameter hasn't been "pre-registered" with it. Good news, I also recently hit this problem and if you rebase to resolve your merge conflict, you will have a new class called `free_form_cmdlinet` which allows turning on arbitrary flags and then passing the full bundle in (introduced in #1913 and used in #1918)

**Review by smowton — APPROVED** [review]
The proper test is a symbol consistency checker, which will have to follow as too busy to write it now. Therefore I say let's merge this and follow up soon.

**Pull request merged**